cpp/dolfin/mesh/cell_types.cpp

```diff
@@ -544,31 +575,88 @@ int mesh::cell_degree(mesh::CellType type, int num_nodes)
   case mesh::CellType::interval:
     return 1;
   case mesh::CellType::triangle:
-    if (num_nodes == 3)
+    switch (num_nodes)
+    {
+    case 3:
       return 1;
-    else if (num_nodes == 6)
+    case 6:
       return 2;
-    else if (num_nodes == 10)
+    case 10:
       return 3;
-    else
-      throw std::runtime_error("Unknown cell type.");
+    case 15:
+      return 4;
+    default:
+      throw std::runtime_error("Triangle order > 4 is not supported.");
+    }
   case mesh::CellType::tetrahedron:
-    if (num_nodes == 4)
-      return 1;
-    else
+    if (num_nodes != 4)
       throw std::runtime_error("Higher order tetrahedron not supported");
+    return 1;
   case mesh::CellType::quadrilateral:
-    if (num_nodes == 4)
-      return 1;
-    else
-      throw std::runtime_error("Higher order quadrilateral not supported");
+  {
+    const int n = std::sqrt(num_nodes);
+    if (num_nodes != n * n)
+    {
+      throw std::runtime_error("Quadrilateral of order "
+                               + std::to_string(num_nodes) + " not supported");
+    }
+    return n - 1;
+  }
   case mesh::CellType::hexahedron:
-    if (num_nodes == 8)
-      return 1;
-    else
+    if (num_nodes != 8)
       throw std::runtime_error("Higher order hexahedron not supported");
+    return 1;
+  default:
+    throw std::runtime_error("Unknown cell type.");
+  }
+}
+//-----------------------------------------------------------------------------
+std::vector<std::uint8_t> mesh::default_cell_permutation(mesh::CellType type,
+                                                         std::int32_t degree)
+{
+  int n;
+  switch (type)
+  {
+  case mesh::CellType::quadrilateral:
+    switch (degree)
+    {
+    case 1:
+      // First order quadrilateral cells does not follow counter
+      // clockwise order (cc), but lexiographic order (LG). This breaks
+      // the assumptions that the cell permutation is the same as the
+      // VTK-map.
+      return {0, 1, 2, 3};
+    default:
+      // Higher order assumes VTK
```

Comment:
This is weird. `Q` elemnets coming from FIAT/FFC should have all lexicographic ordering, why do you take LG for linear but VTK for higher?

Reply:
Because of the software I have available. Gmsh supplies me with higher order meshes, like quads. This will probably be rewritten as I’m working on a permutation function to transfer all VTK input to FEniCS ordering. See one of the other comments.:)

Reply:
I do not follow. What software you have available that generates higher order quad meshes with lexicographic ordering? Moreover, the implementation in dolfin should not depends on what spftware you have available...

Reply:
It’s the opposite way around. The built in meshes in dolfin are ordered lexiographically, while meshes from external sources are ordered VTK style. (Thus also first order quads, hexes). With introducing custom permutation, dolfin supports any consistent ordering. However, we need to assume some defaults for each element type (until I’ve added an actual perturbation function)

Reply:
Therefore I ask, why do you assume default VTK for linear, but LG for other? Why not VTK for all?
The problem is, that you call gmsh for linear quads, but your manual meshes for higher order quads are LG.

Moreover, specification of a permutation could be confusing. Is it a permutation that will be applied (stored) after reading, or is it a permutation with respect to what we in dolfin expect as a local ordering (wrt. FEniCS ordering) (they are in inverse relation)?

Reply:
To not break the built in meshes. I will hopefully have a better solution to this later today. The new idea is to apply the perturbation to the connectivity, changing the layout of the cell to FENICS format.

Reply:
We could assume VTK ordering for them as well, and set a custom_permutation internally.

Reply:
We could assume VTK ordering for them as well, and set a custom_permutation internally.